A space in the path of your globally installed packages causes plugin validation to fail

A space in the path for you globally installed packages causes plugin validation to fail.

If an 'npm config get prefix' command returns a path with a space (like C:\Program Files\nodejs), then a command like 'zowe plugins install @brightside/ops@next' will fail with an error like the following:

```
Command failed: node <C:\Program Files\nodejs\node_modules\@brightside\core\lib\main.js> plugins validate @brightside/ops --response-format-json
internal/modules/cjs/loader.js:582
    throw err;
    ^
Error: Cannot find module 'C:\Program'
```

--- a/__tests__/src/packages/imperative/plugins/suites/InstallingPlugins.ts
+++ b/__tests__/src/packages/imperative/plugins/suites/InstallingPlugins.ts
@@ -229,7 +229,6 @@ describe("Installing Plugins", () => {
     expect(result.stdout).not.toContain(plugins.normal.usage);
 
     // Now go ahead and install the sample
-    console.log(plugins.space_in_path.location);
     result = T.executeTestCLICommand(cliBin, this, [pluginGroup, "install", plugins.space_in_path.location]);
     expect(result.stderr).toMatch(/npm.*WARN/);
     expect(result.stderr).toContain("requires a peer of @brightside/imperative");

--- a/__tests__/src/packages/imperative/plugins/suites/InstallingPlugins.ts
+++ b/__tests__/src/packages/imperative/plugins/suites/InstallingPlugins.ts
@@ -69,6 +69,11 @@ describe("Installing Plugins", () => {
       name    : "normal-plugin-3",
       usage   : "normal-plugin-3"
     },
+    space_in_path: {
+      location: join(__dirname, "../", "test_plugins", "space in path plugin"),
+      name    : "space-in-path-plugin",
+      usage   : "space-in-path-plugin"
+    },
     registry: {
       location: "imperative-sample-plugin",
       name    : "imperative-sample-plugin",
@@ -89,8 +94,8 @@ describe("Installing Plugins", () => {
    * @param {string} cmd     The command to execute on the Test CLI
    * @returns {SpawnSyncReturns<string>} The result of the command execution
    */
-  const executeCommandString = (context: any, cmd: string): SpawnSyncReturns<string> =>
-    T.executeTestCLICommand(cliBin, context, cmd.split(" "));
+  const executeCommandString = (context: any, cmd: string): SpawnSyncReturns<string> => 
+    T.executeTestCLICommand(cliBin, context, cmd.split(" "))
 
   /**
    * The registry from the user's environment, which is used when an explicit registry is not supplied.
@@ -213,6 +218,32 @@ describe("Installing Plugins", () => {
 
     result = executeCommandString(this, "--help");
     expect(result.stdout).toEqual(afterInstall.stdout);
+  });
+
+  it("should install a plugin from a file location that contain space in it path", function(){
+
+    let result = executeCommandString(this, "--help");
+
+    // Verify that the sample plugin isn't there
+    expect(result.stderr).toEqual("");
+    expect(result.stdout).not.toContain(plugins.normal.usage);
+
+    // Now go ahead and install the sample
+    console.log(plugins.space_in_path.location);
+    result = T.executeTestCLICommand(cliBin, this, [pluginGroup, "install", plugins.space_in_path.location]);
+    expect(result.stderr).toMatch(/npm.*WARN/);
+    expect(result.stderr).toContain("requires a peer of @brightside/imperative");
+    expect(result.stderr).toContain("You must install peer dependencies yourself");
+
+    const strippedOutput = T.stripNewLines(result.stdout);
+    expect(strippedOutput).toContain("Registry = " + envNpmRegistry);
+    expect(strippedOutput).toContain(`Installed plugin name = '${plugins.space_in_path.name}'`);
+
+    // Now verify that it got added to the tree
+    result = executeCommandString(this, "--help");
+
+    expect(result.stderr).toEqual("");
+    expect(result.stdout).toContain(plugins.space_in_path.usage);
   });
 
   /* Again we purposely commented out this test because versioning uses a registry,

--- a/__tests__/src/packages/imperative/plugins/suites/ValidatePlugin.ts
+++ b/__tests__/src/packages/imperative/plugins/suites/ValidatePlugin.ts
@@ -35,6 +35,18 @@ describe("Validate plugin", () => {
             expect(result.stdout).toContain("successfully validated.");
         });
 
+        it("when plugin contain space in path is installed sucessfully", () => {
+            const testPlugin = join(testPluginDir, "space in path plugin");
+            const pluginName: string = "space-in-path-plugin";
+            let result = T.executeTestCLICommand(cliBin, this, ["plugins", "install", testPlugin]);
+            expect(result.stdout).toContain(`Installed plugin name = '${pluginName}'`);
+
+            const cmd = `plugins validate`;
+            result = T.executeTestCLICommand(cliBin, this, cmd.split(" "));
+            expect(result.stderr).toEqual("");
+            expect(result.stdout).toContain("successfully validated.");
+        });
+
         it("when provided plugin name is installed successfully", () => {
             const pluginName = "normal-plugin";
             const testPlugin = join(testPluginDir, "normal_plugin");

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/index.d.ts
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/index.d.ts
@@ -1,0 +1,12 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+export * from './sample-plugin';

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/index.js
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/index.js
@@ -1,0 +1,7 @@
+"use strict";
+// import { SamplePlugin } from './sample-plugin';
+// import { IPluginMain, ICommandDefinition } from 'imperative';
+// import chalk from 'chalk';
+// import {join} from "path";
+Object.defineProperty(exports, "__esModule", { value: true });
+//# sourceMappingURL=index.js.map

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/bar/bar.handler.d.ts
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/bar/bar.handler.d.ts
@@ -1,0 +1,19 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+import { ICommandHandler, IHandlerParameters } from "@brightside/imperative";
+
+/**
+ * Defining handler to be use for the 'bar' command.
+ */
+export default class BarHandler implements ICommandHandler {
+    process(params: IHandlerParameters): Promise<void>;
+}

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/bar/bar.handler.js
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/bar/bar.handler.js
@@ -1,0 +1,22 @@
+"use strict";
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
+    return new (P || (P = Promise))(function (resolve, reject) {
+        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
+        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
+        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
+        step((generator = generator.apply(thisArg, _arguments || [])).next());
+    });
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+/**
+ * Defining handler to be use for the 'bar' command.
+ */
+class BarHandler {
+    process(params) {
+        return __awaiter(this, void 0, void 0, function* () {
+            params.response.console.log("@TODO Complete this command: bar");
+        });
+    }
+}
+exports.default = BarHandler;
+//# sourceMappingURL=bar.handler.js.map

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/foo/foo.handler.d.ts
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/foo/foo.handler.d.ts
@@ -1,0 +1,15 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+import { ICommandHandler, IHandlerParameters } from "@brightside/imperative";
+export default class FooHandler implements ICommandHandler {
+    process(params: IHandlerParameters): Promise<void>;
+}

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/foo/foo.handler.js
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/cmd/foo/foo.handler.js
@@ -1,0 +1,19 @@
+"use strict";
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
+    return new (P || (P = Promise))(function (resolve, reject) {
+        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
+        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
+        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
+        step((generator = generator.apply(thisArg, _arguments || [])).next());
+    });
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+class FooHandler {
+    process(params) {
+        return __awaiter(this, void 0, void 0, function* () {
+            params.response.console.log("Invoked sample-plugin foo handler");
+        });
+    }
+}
+exports.default = FooHandler;
+//# sourceMappingURL=foo.handler.js.map

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/healthCheck.handler.d.ts
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/healthCheck.handler.d.ts
@@ -1,0 +1,16 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+import { ICommandHandler, IHandlerParameters } from "@brightside/imperative";
+
+export default class HealthCheckHandler implements ICommandHandler {
+    process(params: IHandlerParameters): Promise<void>;
+}

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/healthCheck.handler.js
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/healthCheck.handler.js
@@ -1,0 +1,20 @@
+"use strict";
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
+    return new (P || (P = Promise))(function (resolve, reject) {
+        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
+        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
+        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
+        step((generator = generator.apply(thisArg, _arguments || [])).next());
+    });
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+class HealthCheckHandler {
+    process(params) {
+        return __awaiter(this, void 0, void 0, function* () {
+            console.log("Invoked health check handler");
+            params.response.build();
+        });
+    }
+}
+exports.default = HealthCheckHandler;
+//# sourceMappingURL=healthCheck.handler.js.map

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/index.d.ts
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/index.d.ts
@@ -1,0 +1,14 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+export * from './healthCheck.handler';
+export * from './cmd/foo/foo.handler';
+export * from './cmd/bar/bar.handler';

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/index.js
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/lib/sample-plugin/index.js
@@ -1,0 +1,9 @@
+"use strict";
+function __export(m) {
+    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
+}
+Object.defineProperty(exports, "__esModule", { value: true });
+__export(require("./healthCheck.handler"));
+__export(require("./cmd/foo/foo.handler"));
+__export(require("./cmd/bar/bar.handler"));
+//# sourceMappingURL=index.js.map

--- a/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/package.json
+++ b/__tests__/src/packages/imperative/plugins/test_plugins/space in path plugin/package.json
@@ -1,0 +1,48 @@
+{
+  "name": "space-in-path-plugin",
+  "version": "1.0.1",
+  "description": "Some description",
+  "main": "lib/index.js",
+  "files": [
+    "lib"
+  ],
+  "imperative": {
+    "name": "space-in-path-plugin",
+    "rootCommandDescription": "imperative working test plugin",
+    "pluginBaseCliVersion": "^1.0.0",
+    "pluginHealthCheck": "./lib/sample-plugin/healthCheck.handler",
+    "pluginAliases": [
+      "np",
+      "normalp"
+    ],
+    "pluginSummary": "This summary is completely different from the description!",
+    "definitions": [
+      {
+        "name": "foo",
+        "description": "dummy foo command",
+        "type": "command",
+        "handler": "./lib/sample-plugin/cmd/foo/foo.handler"
+      },
+      {
+        "name": "bar",
+        "description": "dummy bar command",
+        "type": "command",
+        "handler": "./lib/sample-plugin/cmd/bar/bar.handler"
+      }
+    ]
+  },
+  "peerDependencies": {
+    "TestCLI": "1.0.2",
+    "@brightside/imperative": "1.0.0"
+  },
+  "typings": "lib/index.d.ts",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "EPL 2.0",
+  "private": false,
+  "publishConfig": {
+    "registry": "http://imperative-npm-registry:4873"
+  }
+}

--- a/packages/imperative/src/plugins/utilities/runValidatePlugin.ts
+++ b/packages/imperative/src/plugins/utilities/runValidatePlugin.ts
@@ -39,7 +39,7 @@ export function runValidatePlugin(pluginName: string): string {
 
     const impLogger = Logger.getImperativeLogger();
     impLogger.debug(`Running plugin validation command = ${cmdToRun} plugins validate ${pluginName} --response-format-json`);
-    const valOutputJsonTxt = execSync(`${cmdToRun} plugins validate ${pluginName} --response-format-json`, {
+    const valOutputJsonTxt = execSync(`${cmdToRun} plugins validate "${pluginName}" --response-format-json`, {
         cwd: PMFConstants.instance.PMF_ROOT
     });
     const valResultJsonObj = JSON.parse(valOutputJsonTxt.toString());

--- a/packages/imperative/src/plugins/utilities/runValidatePlugin.ts
+++ b/packages/imperative/src/plugins/utilities/runValidatePlugin.ts
@@ -38,7 +38,7 @@ export function runValidatePlugin(pluginName: string): string {
     cmdToRun += " " + cliPgmToRun;
 
     const impLogger = Logger.getImperativeLogger();
-    impLogger.debug(`Running plugin validation command = ${cmdToRun} plugins validate ${pluginName} --response-format-json`);
+    impLogger.debug(`Running plugin validation command = ${cmdToRun} plugins validate "${pluginName}" --response-format-json`);
     const valOutputJsonTxt = execSync(`${cmdToRun} plugins validate "${pluginName}" --response-format-json`, {
         cwd: PMFConstants.instance.PMF_ROOT
     });

